File: implementations/asap.py
from collections import deque

op_map = {'+': 1, '-': 1, '*': 1, '/': 1, 'O': 0}

def print_vec(arr):
    for val in arr:
        print(val, end=" ")
    print()

def run():
    n = int(input())

    nodes = {}
    for i in range(n):
        id, predc, succ, op = input().split()
        id, predc, succ = map(int, [id, predc, succ])

        preds = [int(x) for x in input().split()]
        succs = [int(x) for x in input().split()]

        node = (preds, succs, op)
        nodes[id] = node

    dq = deque([(1, 1)])
    max_time = 0
    result = {}

    while dq:
        current_node_id, current_cost = dq.popleft()
        current_node = nodes[current_node_id]
        current_cost += op_map[current_node[2]]
        max_time = max(max_time, current_cost)

        result[current_node_id] = max(result.get(current_node_id, 0), max_time)
        for neighbour_id in current_node[1]:
            dq.append((neighbour_id, max_time))

    print("Node, Time")
    for node_id in sorted(result):
        print(f"{node_id} {result[node_id]}")

if __name__ == '__main__':
    run()

File: implementations/asap.cpp
#include <iostream>
#include <bits/stdc++.h>

using namespace std;
using namespace std::chrono;

#define ll long long
#define all(v) v.begin(), v.end()

map<char, int> op_map {{'+', 1}, {'-', 1}, {'*', 1}, {'/', 1}, {'O', 0}};

void print_vec(vector<int> arr) {
  for(auto itr = arr.begin(); itr!=arr.end();itr++) {
    cout<<(*itr)<<" ";
  }
  cout<<endl;
}

struct Node {
    int id;
    vector<int> preds;
    vector<int> succs;
    char op;

    Node(int n, vector<int> p, vector<int> s, char optr) {
        id = n;
        preds = p;
        succs = s;
        op = optr;
    }
};

void run() {
  int n;

  cin>>n;

  map<int,Node> nodes;

  for(int i=0;i<n;i++) {
    int id, predc, succ;
    char op;

    cin>>id>>predc>>succ;
    cin>>op;

    vector<int> preds;
    vector<int> succs;

    for(int i=0;i<predc;i++) {
        int input;
        cin>>input;
        preds.push_back(input);        
    }

    for(int i=0;i<succ;i++) {
        int input;
        cin>>input;
        succs.push_back(input);        
    }
    
    auto node = Node(id, preds, succs, op);    
    nodes.insert(make_pair(id, node));
  }

  deque<pair<int, int>> dq;

  dq.push_back(make_pair(0, 0));

  int max_time = 0;
  map<int, int> result;

  while(!dq.empty()) {
    auto current_node = nodes.at(dq.front().first);
    int current_cost = dq.front().second + op_map[current_node.op];
    max_time = max(max_time, current_cost);
    
    result[current_node.id] = max(result[current_node.id], max_time);
    for(auto neighbour: current_node.succs) {
        dq.push_back(make_pair(neighbour, max_time));
    }
    dq.pop_front();
  }

  cout<<"Node, Time"<<endl;
  for(auto elem : result) {
    cout<<elem.first<<" "<<elem.second<<endl;
  }
}

int main() {

  // cin.tie(0);
  // cin.sync_with_stdio(false);
    

  auto start = high_resolution_clock::now();
  
  run();
  
  auto stop = high_resolution_clock::now();
  auto duration = duration_cast<microseconds>(stop - start);

  // cout<<"time: "<<duration.count()/1e6<<endl;
}